internal/testing/uploads_test.go

```diff
@@ -0,0 +1,175 @@
+package testing
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	rhpv2 "go.sia.tech/core/rhp/v2"
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/api"
+	"lukechampine.com/frand"
+)
+
+type blockedReader struct {
+	rs        api.RedundancySettings
+	read      int
+	data      *bytes.Buffer
+	blockChan chan struct{}
+	readChan  chan struct{}
+}
+
+func newBlockedReader(rs api.RedundancySettings) *blockedReader {
+	data := make([]byte, rhpv2.SectorSize*rs.MinShards)
+	frand.Read(data)
+	return &blockedReader{
+		rs:        rs,
+		data:      bytes.NewBuffer(data),
+		blockChan: make(chan struct{}),
+		readChan:  make(chan struct{}),
+	}
+}
+
+func (r *blockedReader) Read(buf []byte) (n int, err error) {
+	select {
+	case <-r.readChan:
+		<-r.blockChan
+	default:
+	}
+
+	n, err = r.data.Read(buf)
+	if r.read += n; n > 0 && r.read == r.rs.MinShards*rhpv2.SectorSize {
```

Comment:
Instead of comparing against `r.rs.MinShards*rhpv2.SectorSize`, I think this helper would have a bit more utility for other tests as well if we passed in the data to `newBlockedReader` and then just compared against the `len` of that data here.